File: lab1_go/containers/singlelist.go
package containers

import "fmt"

// SingleListNode представляет узел односвязного списка
type SingleListNode struct {
	Data string
	Next *SingleListNode
}

// SingleList представляет односвязный список
type SingleList struct {
	Head *SingleListNode
}

// NewSingleList создает новый односвязный список
func NewSingleList() *SingleList {
	return &SingleList{Head: nil}
}

// Add добавляет элемент в конец списка
func (sl *SingleList) Add(value string) {
	newNode := &SingleListNode{Data: value, Next: nil}

	if sl.Head == nil {
		sl.Head = newNode
		return
	}

	current := sl.Head
	for current.Next != nil {
		current = current.Next
	}
	current.Next = newNode
}

// AddHead добавляет элемент в начало списка
func (sl *SingleList) AddHead(value string) {
	newNode := &SingleListNode{Data: value, Next: sl.Head}
	sl.Head = newNode
}

// AddTail добавляет элемент в конец списка
func (sl *SingleList) AddTail(value string) {
	sl.Add(value)
}

// AddBefore добавляет элемент перед указанным значением
func (sl *SingleList) AddBefore(target, value string) error {
	if sl.Head == nil {
		return fmt.Errorf("список пуст")
	}

	if sl.Head.Data == target {
		sl.AddHead(value)
		return nil
	}

	current := sl.Head
	for current.Next != nil && current.Next.Data != target {
		current = current.Next
	}

	if current.Next == nil {
		return fmt.Errorf("элемент '%s' не найден", target)
	}

	newNode := &SingleListNode{Data: value, Next: current.Next}
	current.Next = newNode
	return nil
}

// AddAfter добавляет элемент после указанного значения
func (sl *SingleList) AddAfter(target, value string) error {
	current := sl.Head
	for current != nil && current.Data != target {
		current = current.Next
	}

	if current == nil {
		return fmt.Errorf("элемент '%s' не найден", target)
	}

	newNode := &SingleListNode{Data: value, Next: current.Next}
	current.Next = newNode
	return nil
}

// Remove удаляет элемент из списка по значению
func (sl *SingleList) Remove(value string) error {
	if sl.Head == nil {
		return fmt.Errorf("список пуст")
	}

	if sl.Head.Data == value {
		sl.Head = sl.Head.Next
		return nil
	}

	current := sl.Head
	for current.Next != nil && current.Next.Data != value {
		current = current.Next
	}

	if current.Next == nil {
		return fmt.Errorf("элемент '%s' не найден в списке", value)
	}

	current.Next = current.Next.Next
	return nil
}

// RemoveHead удаляет элемент с головы списка
func (sl *SingleList) RemoveHead() error {
	if sl.Head == nil {
		return fmt.Errorf("список пуст")
	}
	sl.Head = sl.Head.Next
	return nil
}

// RemoveTail удаляет элемент с хвоста списка
func (sl *SingleList) RemoveTail() error {
	if sl.Head == nil {
		return fmt.Errorf("список пуст")
	}

	if sl.Head.Next == nil {
		sl.Head = nil
		return nil
	}

	current := sl.Head
	for current.Next.Next != nil {
		current = current.Next
	}
	current.Next = nil
	return nil
}

// RemoveBefore удаляет элемент перед указанным значением
func (sl *SingleList) RemoveBefore(target string) error {
	if sl.Head == nil {
		return fmt.Errorf("список пуст")
	}

	if sl.Head.Data == target || sl.Head.Next == nil {
		return fmt.Errorf("недостаточно элементов в списке")
	}

	if sl.Head.Next.Data == target {
		sl.Head = sl.Head.Next
		return nil
	}

	prev := sl.Head
	current := prev.Next

	for current.Next != nil {
		if current.Next.Data == target {
			prev.Next = current.Next
			return nil
		}
		prev = current
		current = current.Next
	}
	return fmt.Errorf("элемент '%s' не найден или перед ним нет элемента", target)
}

// RemoveAfter удаляет элемент после указанного значения
func (sl *SingleList) RemoveAfter(target string) error {
	if sl.Head == nil {
		return fmt.Errorf("список пуст")
	}

	current := sl.Head
	for current != nil && current.Next != nil {
		if current.Data == target {
			current.Next = current.Next.Next
			return nil
		}
		current = current.Next
	}

	return fmt.Errorf("элемент '%s' не найден или после него нет элемента", target)
}

// IsIn проверяет наличие элемента в списке
func (sl *SingleList) IsIn(value string) bool {
	current := sl.Head
	for current != nil {
		if current.Data == value {
			return true
		}
		current = current.Next
	}
	return false
}

// FindIndex возвращает индекс элемента в списке (-1 если не найден)
func (sl *SingleList) FindIndex(value string) int {
	current := sl.Head
	index := 0
	for current != nil {
		if current.Data == value {
			return index
		}
		current = current.Next
		index++
	}
	return -1
}

// GetElement возвращает элемент по индексу
func (sl *SingleList) GetElement(index int) (string, error) {
	if index < 0 {
		return "", fmt.Errorf("индекс не может быть отрицательным")
	}

	current := sl.Head
	currentIndex := 0
	for current != nil {
		if currentIndex == index {
			return current.Data, nil
		}
		current = current.Next
		currentIndex++
	}

	return "", fmt.Errorf("индекс %d выходит за границы списка", index)
}

// GetHead возвращает первый элемент (голову)
func (sl *SingleList) GetHead() (string, error) {
	if sl.Head == nil {
		return "", fmt.Errorf("список пуст")
	}
	return sl.Head.Data, nil
}

// GetTail возвращает последний элемент (хвост)
func (sl *SingleList) GetTail() (string, error) {
	if sl.Head == nil {
		return "", fmt.Errorf("список пуст")
	}

	current := sl.Head
	for current.Next != nil {
		current = current.Next
	}
	return current.Data, nil
}

// Size возвращает размер списка
func (sl *SingleList) Size() int {
	size := 0
	current := sl.Head
	for current != nil {
		size++
		current = current.Next
	}
	return size
}

// Read выводит все элементы списка
func (sl *SingleList) Read() {
	current := sl.Head
	for current != nil {
		fmt.Print(current.Data, " ")
		current = current.Next
	}
	fmt.Println()
}

// ToSlice преобразует список в срез строк
func (sl *SingleList) ToSlice() []string {
	result := []string{}
	current := sl.Head
	for current != nil {
		result = append(result, current.Data)
		current = current.Next
	}
	return result
}

// Clear очищает список
func (sl *SingleList) Clear() {
	sl.Head = nil
}

// IsEmpty проверяет, пуст ли список
func (sl *SingleList) IsEmpty() bool {
	return sl.Head == nil
}

File: lab1_cpp/scripts/containers/singleList.cpp
#include "singleList.h"
#include <iostream>
#include <fstream>

using namespace std;

// Конструктор односвязного списка
SingleList::SingleList() : head(nullptr) {}

// Деструктор односвязного списка
SingleList::~SingleList() {
    clearList(*this);
}

// Добавляет элемент в голову списка
void addToHead(SingleList& list, const string& value) {
    SingleListNode* newNode = new SingleListNode{value, list.head};
    list.head = newNode;
}

// Добавляет элемент в хвост списка
void addToTail(SingleList& list, const string& value) {
    SingleListNode* newNode = new SingleListNode{value, nullptr};
    if (list.head == nullptr) {
        list.head = newNode;
        return;
    }
    SingleListNode* current = list.head;
    while (current->next != nullptr) {
        current = current->next;
    }
    current->next = newNode;
}

// Добавляет элемент перед указанным значением
bool addBefore(SingleList& list, const string& target, const string& value) {
    SingleListNode* current = list.head;
    SingleListNode* prev = nullptr;

    while (current != nullptr) {
        if (current->data == target) {
            SingleListNode* newNode = new SingleListNode{value, current};
            if (prev == nullptr) {
                list.head = newNode;
            } else {
                prev->next = newNode;
            }
            return true;
        }
        prev = current;
        current = current->next;
    }
    return false;
}

// Добавляет элемент после указанного значения
bool addAfter(SingleList& list, const string& target, const string& value) {
    SingleListNode* current = list.head;

    while (current != nullptr) {
        if (current->data == target) {
            SingleListNode* newNode = new SingleListNode{value, current->next};
            current->next = newNode;
            return true;
        }
        current = current->next;
    }
    return false;
}

// Удаляет элемент из списка по значению
bool removeFromList(SingleList& list, const string& value) {
    SingleListNode* current = list.head;
    SingleListNode* prev = nullptr;

    while (current != nullptr) {
        if (current->data == value) {
            if (prev == nullptr) {
                list.head = current->next;
            } else {
                prev->next = current->next;
            }
            delete current;
            return true;
        }
        prev = current;
        current = current->next;
    }
    return false;
}

// Удаляет элемент с головы списка
bool removeFromHead(SingleList& list) {
    if (list.head == nullptr) {
        return false;
    }
    SingleListNode* temp = list.head;
    list.head = list.head->next;
    delete temp;
    return true;
}

// Удаляет элемент с хвоста списка
bool removeFromTail(SingleList& list) {
    if (list.head == nullptr) {
        return false;
    }
    if (list.head->next == nullptr) {
        delete list.head;
        list.head = nullptr;
        return true;
    }
    SingleListNode* current = list.head;
    while (current->next->next != nullptr) {
        current = current->next;
    }
    delete current->next;
    current->next = nullptr;
    return true;
}

// Удаляет элемент перед указанным значением
bool removeBefore(SingleList& list, const string& target) {
    if (list.head == nullptr || list.head->next == nullptr) {
        return false;
    }
    // Если второй элемент - это target, удаляем первый
    if (list.head->next->data == target) {
        SingleListNode* temp = list.head;
        list.head = list.head->next;
        delete temp;
        return true;
    }
    
    SingleListNode* current = list.head->next;
    SingleListNode* prevPrev = list.head;
    
    while (current->next != nullptr) {
        if (current->next->data == target) {
            SingleListNode* temp = current;
            prevPrev->next = current->next;
            delete temp;
            return true;
        }
        prevPrev = current;
        current = current->next;
    }
    return false;
}

// Удаляет элемент после указанного значения
bool removeAfter(SingleList& list, const string& target) {
    SingleListNode* current = list.head;
    
    while (current != nullptr && current->next != nullptr) {
        if (current->data == target) {
            SingleListNode* temp = current->next;
            current->next = temp->next;
            delete temp;
            return true;
        }
        current = current->next;
    }
    return false;
}

// Проверяет, содержится ли элемент в списке
bool isInList(const SingleList& list, const string& value) {
    SingleListNode* current = list.head;
    while (current != nullptr) {
        if (current->data == value) {
            return true;
        }
        current = current->next;
    }
    return false;
}

// Находит индекс элемента в списке
int findIndex(const SingleList& list, const string& value) {
    SingleListNode* current = list.head;
    int index = 0;
    while (current != nullptr) {
        if (current->data == value) {
            return index;
        }
        current = current->next;
        index++;
    }
    return -1;  // элемент не найден
}

// Получает элемент по индексу
string getElement(const SingleList& list, int index) {
    if (index < 0) {
        throw out_of_range("Индекс не может быть отрицательным");
    }
    SingleListNode* current = list.head;
    int currentIndex = 0;
    while (current != nullptr) {
        if (currentIndex == index) {
            return current->data;
        }
        current = current->next;
        currentIndex++;
    }
    throw out_of_range("Индекс выходит за границы списка");
}

// Получает первый элемент (голову)
string getHead(const SingleList& list) {
    if (list.head == nullptr) {
        throw runtime_error("Список пуст");
    }
    return list.head->data;
}

// Получает последний элемент (хвост)
string getTail(const SingleList& list) {
    if (list.head == nullptr) {
        throw runtime_error("Список пуст");
    }
    SingleListNode* current = list.head;
    while (current->next != nullptr) {
        current = current->next;
    }
    return current->data;
}

// Получает размер списка
int getSize(const SingleList& list) {
    int size = 0;
    SingleListNode* current = list.head;
    while (current != nullptr) {
        size++;
        current = current->next;
    }
    return size;
}

// Проверяет, пуст ли список
bool isEmpty(const SingleList& list) {
    return list.head == nullptr;
}

// Читает список
void readList(const SingleList& list) {
    SingleListNode* current = list.head;
    while (current != nullptr) {
        cout << current->data << " ";
        current = current->next;
    }
    cout << endl;
}

// Очищает список
void clearList(SingleList& list) {
    while (list.head != nullptr) {
        SingleListNode* temp = list.head;
        list.head = list.head->next;
        delete temp;
    }
}

// Сохраняет односвязный список в файл
void saveListToFile(const SingleList& list, const string& filePath) {
    ofstream outFile(filePath);
    if (!outFile) {
        throw runtime_error("Ошибка: невозможно открыть файл для записи");
    }
    SingleListNode* current = list.head;
    while (current != nullptr) {
        outFile << current->data << endl;
        current = current->next;
    }
    outFile.close();
}

// Загружает односвязный список из файла
void loadListFromFile(SingleList& list, const string& filePath) {
    ifstream inFile(filePath);
    if (!inFile) {
        throw runtime_error("Ошибка: невозможно открыть файл для чтения");
    }
    clearList(list);
    string value;
    while (getline(inFile, value)) {
        addToTail(list, value);
    }
    inFile.close();
}

// Получает значения из списка
void readSingleList(const SingleList& list) {
    SingleListNode* current = list.head;
    while (current != nullptr) {
        cout << current->data << " ";
        current = current->next;
    }
    cout << endl;
}

// Добавляет значение в список
void addToList(SingleList& list, const string& value) {
    SingleListNode* newNode = new SingleListNode{value, nullptr};
    if (list.head == nullptr) {
        list.head = newNode;
    } else {
        SingleListNode* current = list.head;
        while (current->next != nullptr) {
            current = current->next;
        }
        current->next = newNode;
    }
}